torch/csrc/jit/codegen/cuda/scheduler.cpp

```diff
@@ -142,12 +142,134 @@ constexpr int lastPow2(int n) {
   return std::max(1, n - (n >> 1));
 }
 
+int log2_ceil(int value) {
+  int log2_value = 0;
+  while ((1 << log2_value) < value) {
+    ++log2_value;
+  }
+  return log2_value;
+}
+
+ReductionParams multipleReductionHeuristic(
+    int64_t reduction_dim_size,
+    int64_t outer_dim_size,
+    int64_t inner_dim_size,
+    bool fastest_dim_reduction) {
+  if (fastest_dim_reduction) {
+    TORCH_INTERNAL_ASSERT(reduction_dim_size > 0 && outer_dim_size > 0);
+  } else {
+    TORCH_INTERNAL_ASSERT(
+        reduction_dim_size > 0 && (outer_dim_size > 0 || inner_dim_size > 0));
+  }
+
+  const int64_t kMaxThreadsPerCTA =
+      at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock;
+
+  const int64_t kBlockThresholdNotFastestDim = 64;
+  const int64_t kBlockThresholdFastestDim = 512;
+
+  int gdimx = LaunchParams::UNINITIALIZED_VAL;
+  int gdimy = LaunchParams::UNINITIALIZED_VAL;
+  int bdimx = LaunchParams::UNINITIALIZED_VAL;
+  int bdimy = LaunchParams::UNINITIALIZED_VAL;
+
+  ReductionParams rparams;
+  rparams.fastest_dim = fastest_dim_reduction;
+  rparams.multiple_reds_per_blk = true;
+  rparams.cross_block = false;
+  rparams.cross_grid = false;
+
+  // Is fastest dimension a reduction dimension?
+  if (rparams.fastest_dim) {
+    // TODO: multiple batches per warp causes layer-norm errors
```

Comment:
I think it is a codegen or launch parameters bug, because it only fails on certain offsets when reducing on the fastest dimension and it doesn't appear with 2-d tensors.

e.g. The schedule and kernel is the same for all these examples
1. [20, 100, 35, 67] pass
2. [20, 100, 35, 68] pass
3. [20, 100, 35, 69] fail
4. [20, 100, 35, 70] fail
4. [20, 100, 35, 71] pass